File: backend/app/api/dean.py
"""
SPEC-016: Approve Transfer Application (Dean's Final Decision)
"""
import uuid
from typing import Optional

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import require_role
from app.domain.enums import UserRole
from app.services.dean_service import DeanOfficeService

router = APIRouter()

_require_dean = require_role(UserRole.DEAN_OFFICE)


class ApproveRequest(BaseModel):
    pass


class RejectRequest(BaseModel):
    rejection_code: str
    note: str = ""


@router.get("/applications")
async def list_applications_for_dean(
    program_id: Optional[uuid.UUID] = None,
    period_id: Optional[uuid.UUID] = None,
    current_user=Depends(_require_dean),
    db: AsyncSession = Depends(get_db),
):
    svc = DeanOfficeService(db)
    apps = await svc.list_applications(program_id, period_id)
    return [
        {
            "id": str(a.id),
            "tracking_number": a.tracking_number,
            "status": a.status.value,
            "program": a.program.name if a.program else None,
            "applicant": f"{a.applicant.user.first_name} {a.applicant.user.last_name}" if a.applicant and a.applicant.user else None,
            "ranking_position": a.ranking_entry.position if a.ranking_entry else None,
            "composite_score": float(a.ranking_entry.composite_score) if a.ranking_entry else None,
            "intibak_status": a.intibak_table.status.value if a.intibak_table else None,
        }
        for a in apps
    ]


@router.get("/applications/{application_id}")
async def get_application_detail(
    application_id: uuid.UUID,
    current_user=Depends(_require_dean),
    db: AsyncSession = Depends(get_db),
):
    svc = DeanOfficeService(db)
    app = await svc.get_application_detail(application_id)
    record = app.academic_record
    return {
        "id": str(app.id),
        "tracking_number": app.tracking_number,
        "status": app.status.value,
        "program": app.program.name if app.program else None,
        "applicant": {
            "name": f"{app.applicant.user.first_name} {app.applicant.user.last_name}" if app.applicant and app.applicant.user else None,
            "email": app.applicant.user.email if app.applicant and app.applicant.user else None,
        } if app.applicant else None,
        "academic_record": {
            "gpa_4": float(record.gpa_4) if record and record.gpa_4 else None,
            "gpa_100": float(record.gpa_100) if record and record.gpa_100 else None,
            "yks_score": float(record.yks_score) if record and record.yks_score else None,
        } if record else None,
        "ranking_entry": {
            "position": app.ranking_entry.position,
            "composite_score": float(app.ranking_entry.composite_score),
            "is_primary": app.ranking_entry.is_primary,
        } if app.ranking_entry else None,
        "intibak_table": {
            "id": str(app.intibak_table.id),
            "status": app.intibak_table.status.value,
        } if app.intibak_table else None,
        "english_review": {
            "approved": app.english_proficiency_review.approved,
            "exam_type": app.english_proficiency_review.exam_type,
            "exam_score": float(app.english_proficiency_review.exam_score) if app.english_proficiency_review.exam_score else None,
        } if app.english_proficiency_review else None,
        "documents": [
            {"id": str(d.id), "doc_type": d.doc_type.value, "status": d.status.value}
            for d in app.documents
        ],
    }


@router.post("/applications/{application_id}/approve")
async def approve_final(
    application_id: uuid.UUID,
    request: Request,
    current_user=Depends(_require_dean),
    db: AsyncSession = Depends(get_db),
):
    ip_address = request.client.host if request.client else "unknown"
    svc = DeanOfficeService(db)
    app = await svc.approve_final(application_id, current_user.id, ip_address)
    return {"status": app.status.value, "message": "Transfer Accepted"}


@router.post("/applications/{application_id}/reject")
async def reject_final(
    application_id: uuid.UUID,
    body: RejectRequest,
    request: Request,
    current_user=Depends(_require_dean),
    db: AsyncSession = Depends(get_db),
):
    ip_address = request.client.host if request.client else "unknown"
    svc = DeanOfficeService(db)
    app = await svc.reject_final(
        application_id, current_user.id, body.rejection_code, body.note, ip_address
    )
    return {"status": app.status.value}

File: frontend/src/api/dean.ts
import axios from 'axios'

const client = axios.create({
  baseURL: '/api/dean',
  withCredentials: true,
  timeout: 15000,
  headers: { 'Content-Type': 'application/json' },
})

let _refreshing: Promise<void> | null = null

client.interceptors.response.use(
  (res) => res,
  async (error) => {
    const original = error.config
    if (error.response?.status === 401 && !original._retry) {
      original._retry = true
      try {
        if (!_refreshing) {
          _refreshing = axios
            .post('/api/auth/refresh', {}, { withCredentials: true })
            .then(() => { _refreshing = null })
            .catch((e) => { _refreshing = null; throw e })
        }
        await _refreshing
        return client(original)
      } catch {
        window.location.href = '/login'
      }
    }
    return Promise.reject(error)
  },
)

// ── Types ───────────────────────────────────────────────────────────────

/** One row in the Dean's Office application list. */
export interface DeanApplicationSummary {
  id: string
  tracking_number: string | null
  /** Raw application lifecycle status (RANKING | ANNOUNCED | REJECTED). */
  status: string
  /** Frontend-friendly badge derived from status: Pending | Approved | Rejected. */
  dean_status: 'Pending' | 'Approved' | 'Rejected' | string
  /** Target department / program at IZTECH. */
  program: string | null
  applicant: string | null
  /** Applicant's current institution (academic_record.institution). */
  current_university: string | null
  /** 4.00-scale GPA from academic_record. */
  gpa: number | null
  /** ISO timestamp the applicant submitted the application. */
  submitted_at: string | null
  ranking_position: number | null
  composite_score: number | null
  intibak_status: string | null
}

export type DeanRejectionCode =
  | 'INSUFFICIENT_ACADEMIC_STANDING'
  | 'QUOTA_LIMIT_REACHED'
  | 'DISCIPLINARY_RECORD'
  | 'UNSUITABLE_PROGRAM_MATCH'
  | 'OTHER'

export interface DeanRejectionCodeOption {
  code: DeanRejectionCode
  label: string
}

export interface DeanRejectAuditLog {
  dean_id: string
  dean_email: string | null
  dean_name: string | null
  action: string
  timestamp: string
  rejection_code: DeanRejectionCode
  rejection_reason: string
  note: string
  ip_address: string
}

export interface DeanRejectResponse {
  status: string
  message: string
  rejection_code: DeanRejectionCode
  rejection_reason: string
  notification_message: string
  audit_log: DeanRejectAuditLog
}

/** Standardized dean rejection reasons (SRS AC-01). */
export const DEAN_REJECTION_OPTIONS: DeanRejectionCodeOption[] = [
  {
    code: 'INSUFFICIENT_ACADEMIC_STANDING',
    label: 'Failure to meet the grade point average requirement.',
  },
  { code: 'QUOTA_LIMIT_REACHED', label: 'Program quota limit has been reached.' },
  { code: 'DISCIPLINARY_RECORD', label: 'Disciplinary record on file.' },
  { code: 'UNSUITABLE_PROGRAM_MATCH', label: 'Unsuitable program match.' },
  { code: 'OTHER', label: 'Other reason (see note).' },
]

// ── API ─────────────────────────────────────────────────────────────────

export async function listDeanApplications(): Promise<DeanApplicationSummary[]> {
  const { data } = await client.get('/applications')
  return data
}

export async function approveDeanApplication(id: string): Promise<{ status: string }> {
  const { data } = await client.post(`/applications/${id}/approve`)
  return data
}

export async function listDeanRejectionCodes(): Promise<DeanRejectionCodeOption[]> {
  const { data } = await client.get<DeanRejectionCodeOption[]>('/rejection-codes')
  return data
}

export async function rejectDeanApplication(
  id: string,
  rejection_code: DeanRejectionCode = 'OTHER',
  note: string = '',
): Promise<DeanRejectResponse> {
  const { data } = await client.post<DeanRejectResponse>(`/applications/${id}/reject`, {
    rejection_code,
    note,
  })
  return data
}

/** Full application bundle for the dean's "View Details" modal. */
export interface DeanApplicationDetail {
  id: string
  tracking_number: string | null
  status: string
  dean_status: 'Pending' | 'Approved' | 'Rejected' | string
  program: string | null
  submitted_at: string | null
  applicant: {
    name: string | null
    email: string | null
    national_id: string | null
  } | null
  academic_record: {
    institution: string | null
    gpa_4: number | null
    gpa_100: number | null
    yks_score: number | null
    credits_completed: number | null
  } | null
  english_review: {
    approved: boolean | null
    exam_type: string | null
    exam_score: number | null
    notes: string | null
    reviewed_at: string | null
    must_take_exam: boolean
    exam_date: string | null
    published_at: string | null
  } | null
  department_evaluations: Array<{
    id: string
    score: number | null
    decision: string | null
    notes: string | null
    evaluated_at: string | null
  }>
  ranking_entry: {
    position: number
    composite_score: number
    is_primary: boolean
  } | null
  intibak_table: { id: string; status: string } | null
  documents: Array<{
    id: string
    doc_type: string
    status: string
    file_name: string | null
    file_size_bytes: number | null
    extracted_data: Record<string, unknown> | null
  }>
}

export async function getDeanApplicationDetail(id: string): Promise<DeanApplicationDetail> {
  const { data } = await client.get(`/applications/${id}`)
  return data
}
